src/engraving/layout/layoutsystem.cpp

```diff
@@ -322,52 +366,42 @@ System* LayoutSystem::collectSystem(const LayoutOptions& options, LayoutContext&
             qreal w = lm->width();
             lm->addSystemTrailer(nm);
             if (lm->trailer()) {
-                lm->computeMinWidth();
+                lm->computeWidth(minTicks, 1);
             }
-            minWidth += lm->width() - w;
+            curSysWidth += lm->width() - w;
         }
     }
 
-    //
-    // stretch incomplete row
-    //
-    qreal rest;
-    if (MScore::noHorizontalStretch) {
-        rest = 0;
-    } else {
-        qreal mw          = system->leftMargin();          // DEBUG
-        qreal totalWeight = 0.0;
-
+    // BRING THE WIDTH OF THE SYSTEM TO THE DESIRED VALUE
+    // Gradient descent method: calls the computeWidth() function with a stretch parameter
+    // proportional to the difference between the current length and the target length.
+    // After few iterations, the length will converge to the target length.
+    qreal newRest = systemWidth - curSysWidth;
+    if (ctx.curMeasure == 0 && ((curSysWidth / systemWidth) <= score->styleD(Sid::lastSystemFillLimit))) {
+        // We do not stretch last system if curSysWidth is <= lastSystemFillLimit
+        newRest = 0;
+    }
+    if (MScore::noHorizontalStretch) { // Debug feature
+        newRest = 0;
+    }
+    qreal stretchCoeff = 1;
+    qreal prevWidth = 0;
+    int iter = 0;
+    while (abs(newRest) > score->spatium() * 0.08 && iter < 200) { // spatium*0.08 is about the width of a note stem
```

Comment:
looks like if I'll change a width of a stem in "Properties" panel or in style we'll not layout this properly?

makes sense to use the value from the current style or make a constant variable at stem.cpp

Reply:
Good one, thanks! :)